`socket hang up` error for a lot of links

I have a documentation repository containing close to 250 markdown files. The rule is working well so far. I am creating a new version of the docs by duplicating all of those markdown files under a separate directory.

When I duplicate those files and try to run the linter rule on all of them (close to 500 files), on my local machine it runs fine. But, on my CI server the linter rule is failing to validate external links. All of the errors are due to `socket hang up`. For example:

```log
  52:19  error  http://schemaspy.sourceforge.net/ is dead. (request to http://schemaspy.sourceforge.net/ failed, reason: socket hang up)  no-dead-link
   7:62   error    https://junit.org/junit5/ is dead. (request to https://junit.org/junit5/ failed, reason: Client network socket disconnected before secure TLS connection was established)  no-dead-link
  22:204  error    http://en.wikipedia.org/wiki/Component_model is dead. (request to http://en.wikipedia.org/wiki/Component_model failed, reason: socket hang up)  no-dead-link
  91:151  error  https://www.ietf.org/rfc/rfc2426.txt is dead. (request to https://www.ietf.org/rfc/rfc2426.txt failed, reason: socket hang up)  no-dead-link
  72:279  error    http://jquery.com/ is dead. (request to http://jquery.com/ failed, reason: socket hang up)    no-dead-link
```

These errors are appearing in the CI server consistently but if I don't have the files duplicated, I don't seem to have this issue.

Is there a limitation or performance issue because of the large amount of files/contents?

```

--- a/package.json
+++ b/package.json
@@ -45,6 +45,8 @@
     "get-url-origin": "^1.0.1",
     "minimatch": "^3.0.4",
     "node-fetch": "^2.6.0",
+    "p-all": "^2.1.0",
+    "p-memoize": "^3.1.0",
     "textlint-rule-helper": "^2.1.1"
   },
   "devDependencies": {

--- a/src/no-dead-link.js
+++ b/src/no-dead-link.js
@@ -5,12 +5,15 @@ import fs from 'fs-extra';
 import minimatch from 'minimatch';
 import { isAbsolute } from 'path';
 import { getURLOrigin } from 'get-url-origin';
+import pMemoize from 'p-memoize';
+import pAll from 'p-all';
 
 const DEFAULT_OPTIONS = {
   checkRelative: true, // {boolean} `false` disables the checks for relative URIs.
   baseURI: null, // {String|null} a base URI to resolve relative URIs.
   ignore: [], // {Array<String>} URIs to be skipped from availability checks.
   preferGET: [], // {Array<String>} origins to prefer GET over HEAD.
+  concurrency: 8 // concurrency count of linting link
 };
 
 // Adopted from http://stackoverflow.com/a/3809435/951517
@@ -23,8 +26,9 @@ const URI_REGEXP = /(?:https?:)?\/\/(?:www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}\.[a-z
  */
 function isHttp(uri) {
   const { protocol } = URL.parse(uri);
-  return protocol === "http:" || protocol === "https:"
+  return protocol === 'http:' || protocol === 'https:';
 }
+
 /**
  * Returns `true` if a given URI is relative.
  * @param {string} uri
@@ -87,7 +91,7 @@ async function isAliveURI(uri, method = 'HEAD') {
       'Accept': '*/*',
       // Same host for target url
       // https://github.com/textlint-rule/textlint-rule-no-dead-link/issues/111
-      'Host': host
+      'Host': host,
     },
     // Use `manual` redirect behaviour to get HTTP redirect status code
     // and see what kind of redirect is occurring
@@ -156,7 +160,10 @@ function reporter(context, options = {}) {
   const { Syntax, getSource, report, RuleError, fixer, getFilePath } = context;
   const helper = new RuleHelper(context);
   const opts = Object.assign({}, DEFAULT_OPTIONS, options);
-
+  // 30sec cache
+  const memorizedIsAliveURI = pMemoize(isAliveURI, {
+    maxAge: 30 * 1000,
+  });
   /**
    * Checks a given URI's availability and report if it is dead.
    * @param {TextLintNode} node TextLintNode the URI belongs to.
@@ -202,7 +209,7 @@ function reporter(context, options = {}) {
 
     const result = isLocal(uri)
       ? await isAliveLocalFile(uri)
-      : await isAliveURI(uri, method);
+      : await memorizedIsAliveURI(uri, method);
     const { ok, redirected, redirectTo, message } = result;
 
     if (!ok) {
@@ -269,7 +276,10 @@ function reporter(context, options = {}) {
     },
 
     [`${context.Syntax.Document}:exit`]() {
-      return Promise.all(URIs.map((item) => lint(item)));
+      const linkTasks = URIs.map((item) => () => lint(item));
+      return pAll(linkTasks, {
+        concurrency: opts.concurrency
+      });
     },
   };
 }

--- a/test/no-dead-link.js
+++ b/test/no-dead-link.js
@@ -12,6 +12,8 @@ tester.run('no-dead-link', rule, {
     'should ignore non-http url [ftp](ftp://example.com) by default',
     'should ignore non-http url [websockets](ws://example.com) by default',
     'should be able to check a link in Markdown: [example](https://example.com/)',
+    'should be able to check a link in Markdown: [example](https://jcp.org/en/jsr/detail?id=330)',
+    'should be able to check a link in Markdown: [example](https://dev.mysql.com/downloads/mysql/)',
     'should be able to check a URL in Markdown: https://example.com/',
     'should success with retrying on error: [npm results for textlint](https://www.npmjs.com/search?q=textlint)',
     'should treat 200 OK as alive: https://httpstat.us/200',

